src/dune_rules/expander0.ml

```diff
@@ -11,3 +11,16 @@ let isn't_allowed_in_this_position ~(source : Dune_lang.Template.Pform.t) =
   in
   raise (User_error.E exn)
 ;;
+
+let as_in_build_dir ~what ~loc p =
+  match Path.as_in_build_dir p with
+  | Some p -> p
+  | None ->
+    User_error.raise
+      ~loc
+      [ Pp.textf
+          "%s %s is outside the build directory. This is not allowed."
```

Comment:
So the error will say:
```
susbstitute foo is outside the build directory. This is not allowed.
```
Perhaps it would read better if it said "sustitute: foo is outside the build directory".

Reply:
It's inconsistent with the error messages. This function should be modified if the error message isn't good. Although it seems good enough for me.